inputremapper/injection/consumer_control.py

```diff
@@ -93,23 +104,40 @@ async def run(self):
                 # won't appear, no need to forward or map them.
                 continue
 
+            for listener in self.context.listeners:
+                asyncio.create_task(listener(event))
+                # allow the just created task to do some stuff
+                await asyncio.sleep(0)
+
+            tasks = []
+            results = []
+            if (event.type, event.code) in self.context.callbacks.keys():
+                for callback in self.context.callbacks[(event.type, event.code)]:
+                    ev = copy_event(event)
+                    coroutine = callback(
+                        ev, source=self._source, forward=self._forward_to
+                    )
+                    tasks.append(coroutine)
+                results = await asyncio.gather(*tasks)
+
+            if True in results:
+                continue
+
+            # try legacy injection if the new injection did not do anything
+            # TODO: Remove. only keep the call to forward
```

Comment:
By using pythons for/break/else logic the boolean handled can be avoided:

```py
            for consumer in self._consumers:
                # copy so that the consumer doesn't screw this up for
                # all other future consumers
                event_copy = copy_event(event)
                if consumer.is_handled(event_copy):
                    await consumer.notify(event_copy)
                    break
            else:
                # forward the rest
                if event.type == evdev.ecodes.EV_KEY:
                    logger.debug_key((event.type, event.code, event.value), "forwarding")
                self._forward_to.write(event.type, event.code, event.value)
                # this already includes SYN events, so need to syn here again
```